django/utils/url_config.py

```diff
@@ -0,0 +1,135 @@
+
+BaseParsedURL = namedtuple('ParsedURL', 'scheme username password hostname port path options')
+
+
+class ParsedURL(BaseParsedURL):
+    @property
+    def netloc(self):
+        result = self.hostname
+        if self.port:
+            result += ':{0}'.format(self.port)
+        return result
+
+
+def parse_url(url):
+    """
+    A method to parse URLs into components that handles quirks
+    with the stdlib urlparse, such as lower-cased hostnames.
+    Also parses querystrings into typed components.
+    """
+    if isinstance(url, ParsedURL):
+        return url
+    elif isinstance(url, parse.ParseResult):
+        parsed = url
+    else:
+        parsed = parse.urlparse(url)
+
+    # parsed.hostname always returns a lower-cased hostname
+    # this isn't correct if hostname is a file path, so use '_hostinfo'
+    # to get the actual host
+    hostname, port = parsed._hostinfo
+
+    query = parse.parse_qs(parsed.query)
+
+    options = {}
+    for key, values in query.items():
+        value = values[-1]
+        if value.isdigit():
+            value = int(value)
+        elif value.lower() == 'true':
+            value = True
+        elif value.lower() == 'false':
+            value = False
+
+        options[key] = value
+
+    path = parsed.path[1:]
+
+    return ParsedURL(
+        parsed.scheme,
+        parsed.username,
+        parsed.password,
+        hostname,
+        str(port),
+        path,
+        options
+    )
+
+
+def _register_backend(backend_type, scheme, path):
+    if scheme in BACKENDS[backend_type]:
+        raise ImproperlyConfigured('Scheme {0} has already been registered as a {1} handler'.format(scheme,
+                                                                                                    backend_type))
+    BACKENDS[backend_type][scheme] = path
+
+
+def _get_backend(backend_type, scheme):
+    if scheme not in BACKENDS[backend_type]:
+        raise ImproperlyConfigured('No {0} handler configured for {1}'.format(backend_type, scheme))
+    return BACKENDS[backend_type][scheme]
+
+
+def configure(backend_type, value):
+    scheme = value.split(':', 1)[0]
+
+    backend_path = _get_backend(backend_type, scheme)
+    handler = module_loading.import_string(backend_path)
+
+    if not hasattr(handler, 'config_from_url'):
+        raise TypeError('{0} has no config_from_url method'.format(backend_path))
+
+    if not inspect.ismethod(handler.config_from_url):
+        raise TypeError('{0} is not a class method'.format(handler.config_from_url))
+
+    if backend_path.endswith('.base.DatabaseWrapper'):
+        backend_path = backend_path.replace('.base.DatabaseWrapper', '')
+
+    return handler.config_from_url(backend_path, scheme, value)
+
+
+# Convenience functions
+
+
+def configure_db(value):
+    return configure('db', value)
+
+
+def configure_cache(value):
+    return configure('cache', value)
+
+
+def register_db_backend(scheme, path):
+    if not path.endswith('.base.DatabaseWrapper'):
```

Comment:
I can't find the documentation for how to make a database or cache backend. I'm guessing this is not the only place that explicitly requires this path, but if I'm wrong, should we document this for anyone looking to create their own backends?